x/staking/keeper/delegation.go

```diff
@@ -769,35 +851,95 @@ func (k Keeper) CompleteUnbonding(ctx sdk.Context, delAddr sdk.AccAddress, valAd
 		return nil, err
 	}
 
-	// loop through all the entries and complete unbonding mature entries
+	// loop through all the entries and try to complete unbonding mature entries
 	for i := 0; i < len(ubd.Entries); i++ {
-		entry := ubd.Entries[i]
+		entry := &ubd.Entries[i]
 		if entry.IsMature(ctxTime) {
+			// This hook allows external modules to stop an unbonding delegation entry from fully unbonding
+			// when the completionTime has passed. This allows them to prolong the unbonding period.
+			onHold := k.BeforeUnbondingDelegationEntryComplete(ctx, entry.Id)
+			if onHold {
+				// Mark that the entry is stopped
+				entry.OnHold = true
+			} else {
+				// Proceed with unbonding
+				ubd.RemoveEntry(int64(i))
+				i--
+
+				// track undelegation only when remaining or truncated shares are non-zero
+				if !entry.Balance.IsZero() {
+					amt := sdk.NewCoin(bondDenom, entry.Balance)
+					if err := k.bankKeeper.UndelegateCoinsFromModuleToAccount(
+						ctx, types.NotBondedPoolName, delegatorAddress, sdk.NewCoins(amt),
+					); err != nil {
+						return nil, err
+					}
+
+					balances = balances.Add(amt)
+				}
+			}
+		}
+	}
+
+	// set the unbonding delegation or remove it if there are no more entries
+	if len(ubd.Entries) == 0 {
+		k.RemoveUnbondingDelegation(ctx, ubd)
+	} else {
+		k.SetUnbondingDelegation(ctx, ubd)
+	}
+
+	return balances, nil
+}
+
+// This can be called to complete the unbonding of an unbonding delegation entry that was previously
+// stopped by the BeforeUnbondingDelegationEntryComplete hook in CompleteUnbonding
+func (k Keeper) CompleteStoppedUnbonding(ctx sdk.Context, id uint64) (found bool, err error) {
```

Comment:
I don't think it makes sense to return `bool, err` here. I would just return an `error` (error if the id is not found) and simplify the API